lib/pos_blockchain.go

```diff
@@ -337,15 +333,29 @@ func (bc *Blockchain) validateTimeoutQC(desoBlock *MsgDeSoBlock, validatorSet []
 // validateAncestorsExist checks that all ancestors of this block exist in the block index.
 // If an ancestor is not found, we'll return the block hash of the missing ancestor so the
 // caller can request this block.
-func (bc *Blockchain) validateAncestorsExist(desoBlock *MsgDeSoBlock) (_missingBlockHash *BlockHash, _err error) {
+func (bc *Blockchain) validateAncestorsExist(parentHash BlockHash) *BlockHash {
 	// Notes: starting from the block passed in, we'll look for the parent in the block index.
 	// 1. If the parent does not appear in the block index, we'll return the parent's hash.
 	// 2. If the parent exists in the block index AND is in the best chain, we can safely assume
 	//    that all ancestors exist in the block index.
 	// 3. If the parent exists in the block index but is not in the best chain, we repeat from
 	//    step 1 with the parent as the block passed in.
-	// TODO: Implement me
-	return nil, errors.New("IMPLEMENT ME")
+
+	// check for parent in block index
+	// if parent does not exist in block index, return parent's hash
+	parentBlockNode, exists := bc.blockIndex[parentHash]
+	if !exists {
+		// TODO: should this return an error?
+		return &parentHash
+	}
+
+	// if parent exists in block index AND is in best chain, return nil. We can
+	// safely assume that all ancestor exist then.
+	_, bestChainExists := bc.bestChainMap[parentHash]
+	if bestChainExists {
+		return nil
+	}
+	return bc.validateAncestorsExist(*parentBlockNode.Header.PrevBlockHash)
```

Comment:
Why do we check if the parent is in the best chain? The parent's existence in the block index would guarantee that all of the block's ancestors are known, no?

Reply:
I think I wrote this before we concretified the rules around the block index. I'll update this since we won't be putting anything in the block index unless we have its parent. If it's parent isn't in the block index, it's an orphan and we can figure out how to proceed those in a separate PR